# This is Bit different compared to Standardizers. Let's jot down
1. This works on ROWS unlike standardizers which work on columns
2. This is useful with sparse matrices like text classification
3. The main aim of this fn is to make the" l1 or l2 distance a unit which you will see

# This doesn't require fit because we will always talk in rows when providing test data, and this also works with rows

In [2]:
from sklearn import preprocessing

In [1]:
arr = [[ 1, -1,  2], [ 2,  0,  0], [ 0,  1, -1]]

In [3]:
# Applying the l2 norm and discription will come in below cell
l2_normalized = preprocessing.normalize(arr, norm = 'l2')
l2_normalized

array([[ 0.40824829, -0.40824829,  0.81649658],
       [ 1.        ,  0.        ,  0.        ],
       [ 0.        ,  0.70710678, -0.70710678]])

# Let's take the first row and break it
[1, -1, 2] This resulted in [0.4082..., -0.40882..., 0.8164..] 

If we take sqrt((0.4082)Pow2 + (-0.40882)Pow2 + (0.8164)Pow2) it will result in 1.
I will prove with code in next cell.

In [4]:
import math
res = 0
for i in l2_normalized[0]:
    res += i*i
final = math.sqrt(res)
final
# Got me :)

1.0

In [5]:
# Applying the l1 norm and discription will come in below cell
l1_normalized = preprocessing.normalize(arr, norm = 'l1')
l1_normalized

array([[ 0.25, -0.25,  0.5 ],
       [ 1.  ,  0.  ,  0.  ],
       [ 0.  ,  0.5 , -0.5 ]])

# Let's take the first row and break it
[1, -1, 2] This resulted in [ 0.25, -0.25,  0.5 ]

If we take |0.25| + |-0.25| + |0.5| it will result in 1.

# Hope you understood differece b/w Normalizer and Standardizer. On top of that you saw different Norms.

If you want to see diferent standardizers follow the link below